src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,157 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import breeze.linalg.max
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{MapParam, Param, ParamMap}
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.functions.udf
+import spray.json.DefaultJsonProtocol._
+
+class Trie(trieMap: Map[String, Trie] = Map[String, Trie](),
+           textValue: String = "",
+           normFunc: String => String = (text:String) => text.toLowerCase()) extends Serializable {
+  val value: String = textValue
+
+  def getNodeValue(key: String, value:String): String = (key, value) match {
+    case (k, v) if k.length() == 1 => v
+    case (_, _) => ""
+  }
+
+  def getNode(key: String, value:String): Trie = {
+    val newLeaf = new Trie(textValue = getNodeValue(key, value), normFunc = this.normFunc)
+    this.trieMap.getOrElse(key.substring(0, 1), newLeaf)
+  }
+
+  def put(key:String, value:String, isNormalized:Boolean = false): Trie = (key, value) match {
+    case (_, _) if key.length() == 0 => this.copy()
+    case (k, v) =>
+      var normalizedKey:String = k
+      if (!isNormalized) normalizedKey = this.normFunc(k)
+      val next = this.getNode(normalizedKey, v).put(normalizedKey.substring(1), v, isNormalized = true)
+      val newMap = this.trieMap + (normalizedKey.substring(0, 1) -> next)
+      new Trie(newMap, this.value, this.normFunc)
+  }
+
+  def get(letter:String): Option[Trie] = this.trieMap.get(letter)
+
+  def putAll(stringMap: Map[String, String]): Trie = {
+    var out = this
+    stringMap.foreach {case (k, v) => out = out.put(k, v) }
+    out
+  }
+
+  def copy(): Trie = new Trie(this.trieMap, this.value, this.normFunc)
+
+  def searchForWord(inputText:String, index: Int): (Int, String) = {
+    val normText = this.normFunc(inputText)
+    var current = this.get(normText(index).toString)
```

Comment:
challenge you to remove the vars here

Reply:
It may be possible, but I am not clear on what the net benefit would be. It would require rewriting in a style that falls less in line with the problem to be solved. Is this a team preference?

Reply:
Should make it a trie of chars instead of one-length strings